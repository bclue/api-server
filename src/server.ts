import { NestFactory } from '@nestjs/core';
import { ApplicationModule } from './modules/app.module';
import * as express from 'express';
const bodyParser = require('body-parser');

const instance = express();

async function bootstrap() {
    instance.use(bodyParser.urlencoded( {extended: true} ));
    instance.use(bodyParser.json( {limit: '50mb'} ));
    const app = await NestFactory.create(ApplicationModule, instance);
    app.use(bodyParser.urlencoded( {extended: true} ));
    app.use(bodyParser.json( {limit: '50mb'} ));
    await app.listen(3000);
}
bootstrap();

export { instance }
