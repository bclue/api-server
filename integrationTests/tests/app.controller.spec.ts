const request = require("supertest");
import { instance } from "../../src/server"

describe("/", () => {
    it("GET should return 'hello world'", done => {
        return request(instance)
            .get("/")
            .expect(200)
            .end( (err, response) => {
                if (err) throw err;
                expect(response.text).toMatchSnapshot();
                done();
            });
    });
});

describe("/api", () => {

    it("/GET screener -- returns latest screener", done => {
        return request(instance)
            .get("/api/screener/")
            .expect(200)
            .end( (err, res) => {
                if (err) throw err;

                delete res.body.created;

                expect(res.body).toMatchSnapshot();
                done();
            })
    });

    it("/GET programs -- returns all programs", done => {
        return request(instance)
            .get("/api/program/")
            .expect(200)
            .end( (err, res) => {
                if (err) throw err;

                const data = res.body.map( program => {
                    delete program.created;
                    return program;
                }).sort( (a, b) => a.title.localeCompare(b.title));

                expect(data).toMatchSnapshot();
                done();
            })
    });

    it("/GET:guid -- returns the specific program", done => {
        return request(instance)
            .get("/api/program/GiufEa3ziFwfBEgkjI2gIlfNkk")
            .expect(200)
            .end( (err, res) => {
                if (err) throw err;

                delete res.body.created;

                expect(res.body).toMatchSnapshot();
                done();
            })
    });

    it("POST /notification -- returns programs", done => {
        /*
        {"query":{"bool":{"must":[{"term":{"aish":false}},{"term":{"albertaworks":false}},
        {"term":{"spouse":false}},{"term":{"children":true}},{"range":{"numberchildren":{"lte":2}}},
        {"range":{"grossincome":{"lte":36325}}}]}},
        "meta":{"program_guid":"nQXxT1OMhPGbkqkPHc2QTpofRv","id":"m6b1jkRrVWGTxqr33nmh5gExfD"}}

         */

        const data = {
            "aish": false,
            "albertaworks": false,
            "spouse": false,
            "children": true,
            "numberchildren": 1,
            "grossincome": 10000
        };

        return request(instance)
            .post("/api/notification")
            .set('Content-Type', 'application/json')
            .send(data)
            .expect(201)
            .end( (err, response) => {
                if (err) throw err;

                const data = response.body.map( program => {
                    delete program.created;
                    return program;
                }).sort( (a, b) => a.title.localeCompare(b.title));

                expect(data).toMatchSnapshot();
                done()
            })
    })

});

describe("/protected", () => {

    describe("/login", () => {
       it("POST /login will return { created: true }, authentication is handled by reverse proxy", done => {
            return request(instance)
                .get("/protected/login/")
                .expect(200)
                .end( (err, response) => {
                    if (err) throw err;

                    expect(response.body).toMatchSnapshot();
                    done();
                })
       })
    });



    describe("/program -- CRUD with program datastructures", () => {

        it(" GET /program will return all programs and their queries", done => {
            return request(instance)
                .get("/protected/program/")
                .expect(200)
                .end( (err, response) => {
                    if (err) throw err;

                    const extractedProperties = response.body.map( applicationProgramDto => applicationProgramDto.guid);
                    extractedProperties.sort( (a, b) => a.localeCompare(b));

                    expect(extractedProperties).toMatchSnapshot();
                    done();
                })


        });

        it("GET /program/:guid will return a single program and it's queries", done => {
            return request(instance)
                .get("/protected/program/hp5mupqcVY8ZMiKg7Q91Uoi4Wf")
                .expect(200)
                .end( (err, response) => {
                    if (err)  {
                        console.error(err);
                        throw err;
                    }

                    const applicationProgramDto = response.body;

                    const extractedProps = {
                        guid: applicationProgramDto.guid,
                        user: applicationProgramDto.user,
                        queryIDs: applicationProgramDto.application.map(query => query.id)
                    };

                    extractedProps.queryIDs.sort( (a, b) => a.localeCompare(b));
                    expect(applicationProgramDto.application.some(query => query.guid !== extractedProps.guid)).toBe(false);
                    expect(applicationProgramDto.application.length).toBe(10);
                    done();
                })
        });


        describe("POST /program", () => {
            const guid = "fake_guid";
            const id = "fake_id";
            const data =  {
                "user": {
                    "created": 0,
                    "description": "test POST /protected/program/",
                    "details": "test POST /protected/program/",
                    "externalLink": "test POST /protected/program/",
                    "guid": guid,
                    "tags": [
                        "test POST /protected/program/"
                    ],
                    "title": "test POST /protected/program/"
                },
                "guid": guid,
                "application": [
                    {
                        "guid": guid,
                        "id": id,
                        "conditions": [
                            {
                                "key": {
                                    "name": "empoyment",
                                    "type": "boolean"
                                },
                                "qualifier": "equal",
                                "value": true
                            }
                        ]
                    }
                ]
            };

            it("will create a new program and the queries", done => {
                return request(instance)
                    .post("/protected/program/")
                    .set('Content-Type', 'application/json')
                    .send(data)
                    .expect(201)
                    .end( (err, response) => {
                        if (err)  {
                            console.error(err);
                            throw err;
                        }
                        expect(response.body).toMatchSnapshot();
                        done();
                    })

            });
        });


        describe("PUT /program", () => {
            // copy+paste of data above
            const guid = "fake_guid";
            const id = "fake_id";
            const data =  {
                "user": {
                    "created": 0,
                    "description": "test POST /protected/program/ updated!!!",
                    "details": "test POST /protected/program/",
                    "externalLink": "test POST /protected/program/",
                    "guid": guid,
                    "tags": [
                        "test POST /protected/program/"
                    ],
                    "title": "test POST /protected/program/"
                },
                "guid": guid,
                "application": [
                    {
                        "guid": guid,
                        "id": id,
                        "conditions": [
                            {
                                "key": {
                                    "name": "empoyment",
                                    "type": "boolean"
                                },
                                "qualifier": "equal",
                                "value": true
                            },
                            {
                                "key": {
                                    "name": "age",
                                    "type": "number"
                                },
                                "qualifier": "lessThanOrEqual",
                                "value": 18
                            }
                        ]
                    }
                ]
            };


            it("will update a program data", done => {
                return request(instance)
                    .put("/protected/program")
                    .set('Content-Type', 'application/json')
                    .send(data)
                    .expect(200)
                    .end( (err, res) => {
                        if (err) throw err;

                        expect(res.body).toMatchSnapshot();
                        done();
                    })
            });
        });

        describe("DELETE /program/", () => {
            it("will delete a program and all it's queries", done => {
                return request(instance)
                    .delete("/protected/program/GiufEa3ziFwfBEgkjI2gIlfNkk") // delete "Edmonton Public Library Card"
                    .expect(200)
                    .end( (err, res) => {
                        if (err) throw err;
                        const [ programResponse, queryResponse ] = res.body;
                        expect( programResponse.deleted).toEqual(true);
                        expect( queryResponse.deleted).toEqual(1);
                        done();
                    })
            })
        })
    });

    describe("/query can delete, create or update queries", () => {
        it("can delete a query", done => {
            return request(instance)
                .delete("/protected/query/XsVRvsZeLOLgQhqZLwiLXaHFQk") // delete query from "Optical and Dental Assistance for Seniors"
                .expect(200)
                .end( (err, res) => {
                    if (err) throw err;

                    expect(res.body.deleted).toBe(true);
                    expect(res.body.found).not.toEqual(null);

                    expect(res.body).toMatchSnapshot();
                    done();
                })
        });


        it("can post a new query", done => {
            const data = {
                guid: "fake_guid_2",
                query: {
                    guid: "fake_guid_2",
                    id: "fake_query_id_1",
                    conditions: [
                        {
                            "key": {
                                "name": "empoyment",
                                "type": "boolean"
                            },
                            "qualifier": "equal",
                            "value": true
                        },
                        {
                            "key": {
                                "name": "age",
                                "type": "number"
                            },
                            "qualifier": "lessThanOrEqual",
                            "value": 18
                        }
                    ]
                }
            };

            return request(instance)
                .post("/protected/query")
                .set('Content-Type', 'application/json')
                .send(data)
                .expect(201)
                .end( (err, res) => {
                    if (err) throw err;

                    expect(res.body.created).toEqual(null);
                    expect(res.body.result).toEqual('created');
                    expect(res.body).toMatchSnapshot();
                    done();
                })
        });

        it("can update an existing query", done => {
            const data = {
                guid: "fake_guid_2",
                query: {
                    guid: "fake_guid_2",
                    id: "fake_query_id_1",
                    conditions: [
                        {
                            "key": {
                                "name": "empoyment",
                                "type": "boolean"
                            },
                            "qualifier": "equal",
                            "value": true
                        },
                        {
                            "key": {
                                "name": "age",
                                "type": "number"
                            },
                            "qualifier": "lessThanOrEqual",
                            "value": 18
                        },
                        {
                            "key": {
                                "name": "aish",
                                "type": "boolean"
                            },
                            "qualifier": "equal",
                            "value": true
                        }
                    ]
                }
            };


            return request(instance)
                .post("/protected/query")
                .set('Content-Type', 'application/json')
                .send(data)
                .expect(201)
                .end( (err, res) => {
                    if (err) throw err;

                    expect(res.body.created).toEqual(null);
                    expect(res.body.result).toEqual('updated');
                    expect(res.body).toMatchSnapshot();
                    done();
                })
        })


    });

    describe("/program-description/, can update a user facing program information", () => {
        it("can update an existing program", done => {
            const data = {
                created: 1,
                description: "a program",
                details: "a new detail",
                externalLink: "http://website.com",
                guid: "GiufEa3ziFwfBEgkjI2gIlfNkk", // update Edmonton Library Card
                tags: ["tag1"],
                title: "updated title",
            };


            return request(instance)
                .put("/protected/program-description/")
                .set('Content-Type', 'application/json')
                .send(data)
                .expect(200)
                .end( (err, res) => {
                    if (err) throw err;

                    expect(res.body.created).toEqual(null);
                    expect(res.body.result).toEqual("created");
                    expect(res.body).toMatchSnapshot();
                    done();
                })
        })
    });

    describe("/key/", () => {
        it("can get all keys", done => {
            return request(instance)
                .get("/protected/key/")
                .expect(200)
                .end( (err, res) => {
                    if (err) throw err;

                    expect(res.body).toMatchSnapshot();
                    done();
                })
        });

        it("can create a new key", done => {
            const data = {
                "test_key": {
                    type: "boolean"
                }
            };


            return request(instance)
               .post("/protected/key/")
               .set('Content-Type', 'application/json')
               .send(data)
               .expect(201)
               .end( (err, res) => {
                   if (err) throw err;

                   expect(res.body).toMatchSnapshot();
                   done();
               })
        })
    });

    describe("/screener/", () => {
        it("can get the screener with keys", done => {
            return request(instance)
                .get("/protected/screener/")
                .expect(200)
                .end( (err, res) => {
                    if (err) throw err;

                    delete res.body.created;

                    expect(res.body).toMatchSnapshot();
                    done();
                })
        });

        it("can post a new screener", done => {
           const data = {
               questions: [ "imagine there were questions here" ]
           };

            return request(instance)
                .post("/protected/screener/")
                .set('Content-Type', 'application/json')
                .send(data)
                .expect(201)
                .end( (err, res) => {
                    if (err) throw err;

                    expect(res.body).toMatchSnapshot();
                    done();
                })
        })
    })

});
