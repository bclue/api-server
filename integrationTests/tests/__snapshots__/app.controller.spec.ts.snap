// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/ GET should return 'hello world' 1`] = `"Hello World!"`;

exports[`/protected /login POST /login will return { created: true }, authentication is handled by reverse proxy 1`] = `
Object {
  "created": true,
}
`;

exports[`/protected /program -- CRUD with program datastructures  GET /program will return all programs and their queries 1`] = `
Array [
  "1iXreywnofiF2jKKknlbKRXAtX",
  "1lKYPN8nNaAMldGufGZNEwfjIK",
  "5b1J8A2fhE9cwVo9W5MfbLmuNr",
  "7Vo9xE20WJJJ1ZORDlog1odyzo",
  "8Lqu2z4WCAP9w5sjhcAeOKuHN5",
  "8U2rl4Ha5GXBVqjNStudvzHIch",
  "9dHBCUXa0z8tEZg6QcYMBJVXjf",
  "9rRXexrA3JzmXbFdKCmjtTj57l",
  "BcKJhMTErAsHfXq1jzealos8kp",
  "bzf4ryrVVsKkqpsEA5dBZd2kP5",
  "EY1QV5T6AyB4Ykl2Lt1Nt9kf9g",
  "GiufEa3ziFwfBEgkjI2gIlfNkk",
  "HiGYI2bhBoQ6n82R9ptoNZ0oaQ",
  "hp5mupqcVY8ZMiKg7Q91Uoi4Wf",
  "iHPbkwKFsck9ZLnMePKSj5e5mz",
  "IO2YsSRkLCIZwIcYGjO6rNkxUE",
  "JIcpL1cap5Eg0ikPk2tecp6JG7",
  "K08iJHYVRGL5pBOomOvwa2zwX2",
  "kc0ov01FtBgr88C5P6EE6lnC8s",
  "NMh2Fu1IG5XnpOf5b6zu2t1Hh6",
  "nQXxT1OMhPGbkqkPHc2QTpofRv",
  "NTVwRTvMb82TIB5VLbf3v36tZN",
  "OGR28icdrofstoYETvIGg19XAv",
  "oZ68OrZFpvkMMGHqV3CFELhuVY",
  "sgKMX0336kgepHbPeObRLOvcFL",
  "UbdhHxTUe3N5zm5BxR6WQ2i6vf",
  "Y5KEzulkwxviOuzB4Ny5QTzMGD",
  "zXRcJxGzIrwhmk5LTbE4jdwBMh",
]
`;

exports[`/protected /program -- CRUD with program datastructures GET /program/:guid will return a single program and it's queries 1`] = `
Object {
  "guid": "hp5mupqcVY8ZMiKg7Q91Uoi4Wf",
  "queryIDs": Array [
    "C8NZoL1Nf7VEgEC4VMFI9GlebN",
    "d5NswIlUvzxY0AJ1hGm4eKQtS4",
    "fMwdx1jiAZUt09ALDer98GardT",
    "JeXk8cOZnqJMepmHpxDqpsEXNE",
    "jI1HlTlpbULsSVUmHnYi3ZlVJ7",
    "jzMQkEK41gnhmwWyxAuWF3eZ4d",
    "qZyokzKnpIw3Ue6mdTi8je3knK",
    "vCYJSMCl8vxwkYKpU74Gcy1BVg",
    "wgLgrNpNO6WYMjMSbiGiI4ABLt",
  ],
  "user": Object {
    "created": 1511470231394,
    "description": "Extended health coverage for adults with low incomes",
    "detailLinks": Array [],
    "details": "What is it?
The Alberta Adult Health Benefit (AAHB) program supports Albertans with limited incomes by covering the cost of prescription drugs, dental services, optical services, emergency ambulance, and diabetic supplies. The AAHB pays for supplies that are not available through standard Alberta health coverage.

Who is eligable?
Residents of Alberta are eligible for the AAHB if they meet the income criteria as well as one of the following conditions:
- Leaving Alberta Works Income Support or Assured Income for the Severely Handicapped (AISH) and has income from employment, self-employment, or the Canada Pension Plan Disability program
- High ongoing prescription drug needs
- Pregnant

An individual or family is NOT eligible for AAHB if they are receiving extended health benefits from other government programs including: 
- Income Support/AISH
- Child and Youth Supports (Child Health Benefit)
- Government Funded Indigenous Health Programs
- Sponsored Immigrants or Refugees
- Seniors (65+) Health Programs

What do I need to apply?
- SIN numbers for all family members
- Alberta Personal Health Number
- Completed your taxes so that Alberta Health Services can check your income

For more information and to find the application follow the link below",
    "externalLink": "www.alberta.ca/alberta-child-health-benefit.aspx",
    "guid": "hp5mupqcVY8ZMiKg7Q91Uoi4Wf",
    "tags": Array [
      "Health",
      "Alberta",
      "Adult",
    ],
    "title": "Adult Health Benefit",
  },
}
`;

exports[`/protected /program -- CRUD with program datastructures POST /program will create a new program and the queries 1`] = `
Object {
  "created": true,
}
`;

exports[`/protected /program -- CRUD with program datastructures PUT /program will update a program data 1`] = `
Object {
  "updated": true,
}
`;

exports[`/protected /program-description/, can update a user facing program information can update an existing program 1`] = `
Object {
  "created": true,
}
`;

exports[`/protected /query can delete, create or update queries can delete a query 1`] = `
Object {
  "deleted": true,
  "found": true,
}
`;

exports[`/protected /query can delete, create or update queries can post a new query 1`] = `
Object {
  "created": true,
}
`;

exports[`/protected /query can delete, create or update queries can update an existing query 1`] = `
Object {
  "created": true,
}
`;
